### ISSUE TITLE
Support SSR

### ISSUE THREAD

**Issue by cythrawll** [apexcharts/apexcharts.js#3703]
## Summary

SSR is becoming more and more popular way to handle rendering, and having charts that can render both server and client side is really appealing for a number of reasons. While most SSR frameworks allow you to load client only libraries without SSR, apex in theory should be able to be used still client side only. (however in the case of react18s example, react-apexcharts seems to currently be broken as dynamic esm import seems to be broken there is a ticket for that already). However, making libs client side only is kinda a painful work around that isn't apparent when first using the library as docs don't mention anything about this gotcha.

## API Changes

Make apex SSR friendly.

## Intended Use Case

Better DX and UX when using apex charts in SSR frameworks like NEXT and Remix.  Taking advantage of SSR frameworks caching and time to first render as well as other server side rendering benefits. I love using Apex, but this is probably going to cause me to abandon this in favor of other libraries.

**Comment by rakirox** [comment]
+1

**Comment by oleg-savenok** [comment]
+1

**Comment by bf** [comment]
@cythrawll can you share what other libraries are like apexcharts but support SSR with frameworks like nuxt3?

**Comment by cythrawll** [comment]
https://nivo.rocks/

**Comment by vignesh-gupta** [comment]
Guys, I think first @ApexCharts Please reopen the issue. and Any updates on this @bf @cythrawll. Because I'm using Next13 and I've added `"use client"` on top the file but still throws this error in my vs console- 

```
- error node_modules/apexcharts/dist/apexcharts.common.js (6:379433) @ eval
- error ReferenceError: window is not defined
    at __webpack_require__ (/home/vignesh/Desktop/Code it/projects/hcl-form-proto/.next/server/webpack-runtime.js:33:42)
    at __webpack_require__ (/home/vignesh/Desktop/Code it/projects/hcl-form-proto/.next/server/webpack-runtime.js:33:42)
    at eval (./app/survey/result/page.tsx:9:74)
    at (ssr)/./app/survey/result/page.tsx (/home/vignesh/Desktop/Code it/projects/hcl-form-proto/.next/server/app/survey/result/page.js:457:1)
    at __webpack_require__ (/home/vignesh/Desktop/Code it/projects/hcl-form-proto/.next/server/webpack-runtime.js:33:42)
null
```

**Comment by machour** [comment]
Would love to get SSR support on apex

**Comment by codesjedi** [comment]
+1

**Comment by sannajammeh** [comment]
> Guys, I think first @apexcharts Please reopen the issue. and Any updates on this @bf @cythrawll. Because I'm using Next13 and I've added `"use client"` on top the file but still throws this error in my vs console-
> 
> ```
> - error node_modules/apexcharts/dist/apexcharts.common.js (6:379433) @ eval
> - error ReferenceError: window is not defined
>     at __webpack_require__ (/home/vignesh/Desktop/Code it/projects/hcl-form-proto/.next/server/webpack-runtime.js:33:42)
>     at __webpack_require__ (/home/vignesh/Desktop/Code it/projects/hcl-form-proto/.next/server/webpack-runtime.js:33:42)
>     at eval (./app/survey/result/page.tsx:9:74)
>     at (ssr)/./app/survey/result/page.tsx (/home/vignesh/Desktop/Code it/projects/hcl-form-proto/.next/server/app/survey/result/page.js:457:1)
>     at __webpack_require__ (/home/vignesh/Desktop/Code it/projects/hcl-form-proto/.next/server/webpack-runtime.js:33:42)
> null
> ```

Next always renders "use client" on the server too, its just a way to tell Next to ship the JS to the client as well.
You need to use Next.js dynamic import combined with `{ssr: false}`.

**Comment by abdurrahmanseyidoglu** [comment]
Here is my workaround: 
```
<script setup>
const VueApexCharts = ref(null);

onMounted(async () => {
  const {default: ApexCharts} = await import('vue3-apexcharts');
  VueApexCharts.value = ApexCharts;
});
</script>

<template>
   <div v-if="VueApexCharts">
      <VueApexCharts :options="chartOptions" :series="series" height="350" type="area"></VueApexCharts>
    </div>
</template>
```

**Comment by abdulrehmanaxa1** [comment]
It create problem in Angular side as well, Got this issue

5:59:58 PM [vite] Error when evaluating SSR module /chunk-JAU2AQNA.mjs:
|- ReferenceError: window is not defined
    at node_modules/apexcharts/dist/apexcharts.common.js

**Comment by Creative-Difficulty** [comment]
@apexcharts REOPEN

**Comment by daviddavid** [comment]
As @sannajammeh said, you can use apexcharts in NextJS by importing it like this:

```javascript
import dynamic from "next/dynamic";
const Chart = dynamic(() => import("react-apexcharts"), { ssr: false });
```

**Comment by BeSaRa** [comment]
any update for Angular Side ?? how to use it ApexChart with SSR ???? really it is very important feature

**Comment by ts-ben** [comment]
I am able to render and build `react-apexcharts` on the server in one route, but get a build error in another route.

for `app/[[...analytics]]` `react-apexcharts` is rendered on the server. I do get an error, but still able to build and render on the server.
In another route `app/dashboard`, I get a build error and the build crushes. In this route, I have to use 
`const ReactApexCharts = dynamic(() => import('react-apexcharts'))`
It makes the build work, but when I load the page, everything is rendered on the server, but the chart take a second to render, because it is rendered on the client. And it looks not very good.

It's weird because on the route `app/[[...analytics]]` the chart is rendered on the server together with the whole page, and I don't have to use dynamic loading.

What the hell is going on?

**Comment by phil-hudson** [comment]
Any update for React?

**Comment by Creative-Difficulty** [comment]
Why is this even closed if it doesnt work for anyone?????

**Comment by Diyouf** [comment]
Does anyone How can use the apexChart in Remix js

**Comment by BeSaRa** [comment]
any update here please about Angular SSR with Apexcharts

**Comment by kinan-almoalm** [comment]
+1

**Comment by IgnisDa** [comment]
I was using remix and got recharts working like so:

```tsx
import { lazy, Suspense } from "react";

const ApexChart = lazy(async () => {
	const module = await import("react-apexcharts");
	return { default: module.default };
});

const MyChart = () => {
       return 	<Suspense>
				<ApexChart
					series={[
						{
							name: "series-1",
							data: [30, 40, 45, 50, 49, 60, 70, 91],
						},
					]}
					options={{
						chart: {
							id: "apexchart-example",
						},
						xaxis: {
							categories: [
								1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999,
							],
						},
					}}
					type="bar"
					width="500"
				/>
			</Suspense>
}
```

It is not typesafe but it works 🤷🏽‍♂️.

**Comment by TomazDimas** [comment]
+1

**Comment by junedchhipa** [comment]
This has been supported since **v5.10.0**, and nobody came back to say so on this thread. Apologies for that.

SSR landed as `SSRRenderer`, exposed through the `apexcharts/ssr` export condition:

```js
import { SSRRenderer } from 'apexcharts/ssr'

const svgString = await SSRRenderer.renderToString(
  {
    chart: { type: 'bar' },
    series: [{ name: 'S', data: [30, 40, 35, 50] }],
    xaxis: { categories: ['a', 'b', 'c', 'd'] },
  },
  { width: 500, height: 300 },
)
```

`package.json` routes the `node` condition to the SSR build automatically, so Next.js, Nuxt, SvelteKit and plain Node resolve it without extra configuration. Internally all browser access goes through a `BrowserAPIs` shim guarded by `Environment.isSSR()`, which is what removed the `window is not defined` class of failure.

Re-tested on **6.8.0** in plain Node before posting this, no DOM and no jsdom:

```
renderToString -> string, 12195 chars
starts with <svg>: true
apexcharts-bar-area paths: 4
x-axis labels a..d present: true
no window/document errors
```

Worth noting for anyone arriving from a search: the commit that added this closed #4288 and did not reference this issue, which is why this thread stayed quiet and was then auto-closed by the stale bot rather than as done. Leaving it closed, now with the reason recorded. If you hit a specific SSR gap on 6.8.0, please open a fresh issue with the framework and version.
